### PR TITLE
feat(btrfs): support unmounted subvolumes

### PR DESCRIPTION
**Pull request by rogarb**
This option allows the creation of subvolumes which won't be mounted, allowing to create more complex btrfs layouts.

This is a personal attempt to fix the issue #288, since it didn't seem to get much attention.

### CODE REVIEW

**Comment by lilyinstarlight** [comment]
I think it might be better if we reconsider whether to automatically create mounts when the parent btrfs does not have a `mountpoint` and subvolumes also do not `mountpoint` rather than a one-off option on each subvol for this. I kinda regret a bit making it default to adding mountpoints in my initial btrfs subvol PR for that case 🙃

Maybe we can add a btrfs option like `mountSubvolumes` that defaults to `true` but allows disabling that behavior, and then do `config.mountpoint == null && config.mountSubvolumes` in the conditional? Or maybe it would be better to make a backwards incompatible change now rather than later and just force people to add `mountpoint` to each subvolume rather than letting it be implicit if the btrfs filesystem root itself isn't mounted

Do you have thoughts @Lassulus?

**Comment by Lassulus** [comment]
how about we set the default of mountpoint to config.name? people can then set it to null if they don't want it mounted, that should simplify the code also in general

**Comment by rogarb** [comment]
> how about we set the default of mountpoint to config.name? people can then set it to null if they don't want it mounted, that should simplify the code also in general

Great, I will have a look and try to implement that.

**Comment by rogarb** [comment]
This is my attempt to solve the problem, I commented the code to explain assumptions, it is now waiting for review.

**Comment by Lassulus** [comment]
I don't really understand the usecase why we want to automatically mount the subvolumes if the btrfs mainvolume has not mountpoint? 

wouldn't be a default in the subvolume like this better?:
```
 mountpoint = lib.mkOption {
  type = lib.types.nullOr diskoLib.optionTypes.absolute-pathname;
  default = if config.mountpoint == null then null else "${config.mountpoint}/${config.name}";
  description = "Location to mount the subvolume to (set to null if you don't want the subvolume to be mounted).";
};
```

we have to change the args in line 25 a bit, and change it basically to `{ name, ...}@subvol`

**Comment by lilyinstarlight** [comment]
> I don't really understand the usecase why we want to automatically mount the subvolumes if the btrfs mainvolume has not mountpoint?

Yeah that's what I was saying, I regret doing that and would not mind if we just reversed that default entirely

> wouldn't be a default in the subvolume like this better?:

No because that subvolume wouldn't need to be mounted at all. Subvolumes only need to be mounted if their path would be different (e.g. mounting a `/` subvol if it is different from the filesystem's root subvol and then mounting each other subvolume where it goes within the `/` subvol, for my case)

**Comment by Lassulus** [comment]
I'm not sure I understand fully. My idea was: if the subvolume.mountpoint has a value of null we skip mounting it, and the default value is just the root mountpoint + the name. if the root mountpoint has null as a value we skip mounting the subvolume by default.

**Comment by lilyinstarlight** [comment]
> My idea was: if the subvolume.mountpoint has a value of null we skip mounting it, and the default value is just the root mountpoint + the name.

Yeah if you `btrfs subvolume create /mountpoint/subvolname` then you don't need to mount it at `/mountpoint/subvolname`. It already exists there

We should just skip mounting subvols by default entirely if we do not want to do it for the case that the root filesystem is not mounted

**Comment by rogarb** [comment]
> > My idea was: if the subvolume.mountpoint has a value of null we skip mounting it, and the default value is just the root mountpoint + the name.
> 
> Yeah if you `btrfs subvolume create /mountpoint/subvolname` then you don't need to mount it at `/mountpoint/subvolname`. It already exists there
> 
> We should just skip mounting subvols by default entirely if we do not want to do it for the case that the root filesystem is not mounted

So basically, we should only mount {main,sub}volumes if and only they have a `mountpoint` value, right? Since any "sub-subvolume" will appear automatically inside their mounted parent. In that case, it is even better to have  a default value of `null` for the subvolume mountpoints.

I think that it's a bit different from the ZFS behaviour, which is what might have influenced my thoughts when considering use cases.

**Comment by rogarb** [comment]
> I don't really understand the usecase why we want to automatically mount the subvolumes if the btrfs mainvolume has not mountpoint?

Well in my case, I don't mount the main volume at all and I mount different subvolumes to different places (e.g. `SYSTEM/nixos` to `/nix`, `DATA/nixos-config` to `/etc/nixos`, `DATA/home` to `/home`...). You can think about it as a LVM-like use case.

**Comment by rogarb** [comment]
This simple last patch should behave as stated. 

The only case I can think of where you want to actually act on mounting and don't specify a different mountpoint is when you have specific mount options for a subvolume. Maybe we should check `subvol.mountpoint != null || subvol.mountOptions != []`?

**Comment by Lassulus** [comment]
as far as I understood things, this seems correct. waiting for an approval from @lilyinstarlight and will merge then :)

**Review by lilyinstarlight — COMMENTED** [review]
The change looks good to me and I just successfully tested it with my config by adding explicit mountpoints for my subvolumes

Can you update the `examples/btrfs-subvolumes.nix` file to have explicit mountpoints and add a note to `docs/upgrade-guide.md` that subvolume mounts when the parent filesystem has no mount points need to be explicit now?

Thank you for this contribution!

**Comment by rogarb** [comment]
Le 2023-07-14 16:48, Lily Foster a écrit :
> @lilyinstarlight commented on this pull request.
> 
> The change looks good to me and I just successfully tested it with my
> config by adding explicit mountpoints for my subvolumes
> 
> Can you update the examples/btrfs-subvolumes.nix file to have explicit
> mountpoints and add a note to docs/upgrade-guide.md that subvolume
> mounts when the parent filesystem has no mount points need to be
> explicit now?
> 

I will be AFK for a few days, I will take care of all this when I come 
back.

**Comment by rogarb** [comment]
> Can you update the `examples/btrfs-subvolumes.nix` file to have explicit mountpoints and add a note to `docs/upgrade-guide.md` that subvolume mounts when the parent filesystem has no mount points need to be explicit now?

I just added two commits addressing these 2 points. Please have a look and let me know if anything has to be changed.

I also thought about a corner case: if a subvolume has `mountOptions` and no `mountpoint`, **and** its parent is mounted, then it will appear in the filesystem hierarchy but it will not be mounted with the specified options.
In order to address that (if there is something to address which I am not sure), we could check for `mountOptions != []` in `_mount`, but in that case, we should also derive a mountpoint using the mounted parent's mountpoint.
My opinion is that for such case, `mountpoint` should be also specified, and thus no change in code would be needed. In any case, there could be somewhere a warning telling that a subvolume has `mountOptions` without a mountpoint, so the options will not be taken into account.

**Comment by lilyinstarlight** [comment]
> My opinion is that for such a case, `mountpoint` should also be specified, and thus no change in code would be needed.
> 
> In any case, there could be somewhere a warning when a subvolume has `mountOptions` without a`mountpoint`, so the options will not be taken into account.

Yeah, I would agree that `mountpoint` should be specified in that case. Feel free to add a warning if you'd like

I'll review this one more time today or tomorrow, thank you!

**Comment by rogarb** [comment]
> > My opinion is that for such a case, `mountpoint` should also be specified, and thus no change in code would be needed.
> > In any case, there could be somewhere a warning when a subvolume has `mountOptions` without a`mountpoint`, so the options will not be taken into account.
> 
> Yeah, I would agree that `mountpoint` should be specified in that case. Feel free to add a warning if you'd like

I added a new commit with a warning (and rebased to current main branch).

> I'll review this one more time today or tomorrow, thank you!

Great, thank you!

File: lib/types/btrfs.nix
Comment (rogarb) [inline lib/types/btrfs.nix:94]: This actually triggers the warning all the time since mountOptions.default is [ "defaults" ] and not the empty list... 
Is there a way to get the corresponding Option type default value for the comparison here?

Comment (lilyinstarlight) [reply lib/types/btrfs.nix:94]: `options.subvolumes.[name].default` should get it, I think. But it would probably be better to just check if it's set by `options.subvolumes.[name].highestPrio < (mkOptionDefault {}).priority`. That will tell if it's set at all aside from the option's default value, even if the value is set to what happens to be the default

Comment (Lassulus) [reply lib/types/btrfs.nix:94]: Uh I think checking against the default value would be better. Since we reeval the config in the tests and then we could run in weird behavior

Comment (lilyinstarlight) [reply lib/types/btrfs.nix:94]: What weird behavior?

I'm not sure I am able to come up with something weird that would happen (re-eval doesn't matter, this just checks what the highest option priority is set at all which only changes when someone sets it either without a priority or with a higher priority manually)

Fwiw this is how checking whether an option changed from the default is done in nixpkgs a few places

Comment (Lassulus) [reply lib/types/btrfs.nix:94]: well I pass the whole config.disko.devices into config.disko.devices again (of the test target system) so everything would be set to the default value without it still being the default value.

Comment (lilyinstarlight) [reply lib/types/btrfs.nix:94]: Yeah but that won't change the priority value, right? (from the special type `{ _type = "override"; priority = ...; content = ...; }`)

Comment (lilyinstarlight) [reply lib/types/btrfs.nix:94]: Oh you mean the merged config values. Huh, yeah I guess that could do something weird. Yeah I guess comparing against the default is fine then

Comment (rogarb) [reply lib/types/btrfs.nix]: I found a way to access the default value using `(options.subvolumes.type.getSubOptions []).mountOptions.default`, if you think there is a simpler way, please let me know.

**Review by lilyinstarlight — APPROVED** [review]
This diff looks good to me now!

It might be good for @Lassulus to do another sanity check on it, but I don't see any problems and my disko script actually comes out the same before and after this PR now that I added `mountpoint`s to the subvolumes

File: docs/upgrade-guide.md
Comment (tiredofit) [inline docs/upgrade-guide.md:36]: I believe you will want `mountpoint` instead of `mountpount`

Comment (rogarb) [reply docs/upgrade-guide.md]: Thank you for spotting this! I just pushed a fixed revision!

**Review by Lassulus — APPROVED** [review]

**Pull request merged**
